baseclass/src/test/java/com/google/apigee/calloutsbaseclass/tests/CalloutsBaseTest.java

```diff
@@ -0,0 +1,215 @@
+package com.google.apigee.calloutsbaseclass.tests;
+
+import com.apigee.flow.message.MessageContext;
+import com.google.apigee.calloutbaseclass.CalloutsBase;
+import org.junit.Assert;
+import org.junit.Before;
+import org.junit.Test;
+import org.mockito.MockitoAnnotations;
+import org.mockito.Spy;
+
+import java.io.ByteArrayOutputStream;
+import java.io.PrintStream;
+import java.io.PrintWriter;
+import java.io.StringWriter;
+import java.util.HashMap;
+import java.util.Map;
+
+/**
+ * Copyright 2020 Google LLC
+ *
+ * <p>Licensed under the Apache License, Version 2.0 (the "License"); you may not use this file
+ * except in compliance with the License. You may obtain a copy of the License at
+ *
+ * <p>https://www.apache.org/licenses/LICENSE-2.0
+ *
+ * <p>Unless required by applicable law or agreed to in writing, software distributed under the
+ * License is distributed on an "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either
+ * express or implied. See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+public class CalloutsBaseTest {
+
+  private static final String ERROR_FLOW_VARIABLE = "error";
+  private static final String EXCEPTION_FLOW_VARIABLE = "exception";
+  private static final String EXCEPTION_STACKTRACE_FLOW_VARIABLE = "exception_stacktrace";
+  private static final String LOG_FLOW_VARIABLE = "log";
+  private static final String OPTIONAL_VARIABLE_KEY = "optionalVar";
+  private static final String OPTIONAL_VARIABLE_VALUE = "optionalValue";
+  private static final String OPTIONAL_VARIABLE_EMPTY_VALUE_KEY = "optionalVarEmpty";
+  private static final String REQUIRED_VARIABLE_KEY = "requiredVar";
+  private static final String REQUIRED_VARIABLE_VALUE = "requiredValue";
+  private static final String REQUIRED_VARIABLE_EMPTY_VALUE_KEY = "requiredVarEmpty";
+  private static final String TEST_LOG_STATEMENT = "asdf123";
+  private static final String TEST_LOG_STATEMENT2 = "defgh5678";
+
+  static class CalloutsBaseTestImpl extends CalloutsBase {
+    public CalloutsBaseTestImpl(Map properties) {
+      super(properties);
+    }
+  }
+
+  abstract static class FakeMessageContext implements MessageContext {
+    private Map<String, Object> variables;
+
+    private Map<String, Object> getVariables() {
+      if (variables == null) {
+        variables = new HashMap<>();
+      }
+      return variables;
+    }
+
+    @Override
+    public Object getVariable(final String name) {
+      return getVariables().get(name);
+    }
+
+    @Override
+    public boolean setVariable(final String name, final Object value) {
+      getVariables().put(name, value);
+      return true;
+    }
+
+    @Override
+    public boolean removeVariable(final String name) {
+      if (getVariables().containsKey(name)) {
+        variables.remove(name);
+      }
+      return true;
+    }
+  }
+
+  CalloutsBaseTestImpl calloutsBase;
+  @Spy FakeMessageContext messageContext;
```

Comment:
why @Spy?

Reply:
@SPY allows me to implement only the few methods of interest in MessageContext and leave the rest abstract.